lib/graphql/analysis/ast/query_complexity.rb

```diff
@@ -8,113 +8,211 @@ class QueryComplexity < Analyzer
-        end
-
-        # Get a complexity value for a field,
-        # by getting the number or calling its proc
-        def get_complexity(ast_node, field_defn, child_complexity, visitor)
-        # Return if we've visited this response path before (not counting duplicates)
-          defined_complexity = field_defn.complexity
-
-          arguments = visitor.arguments_for(ast_node, field_defn)
-
-          case defined_complexity
-          when Proc
-            defined_complexity.call(visitor.query.context, arguments, child_complexity)
-          when Numeric
-            defined_complexity + (child_complexity || 0)
-          else
-            raise("Invalid complexity: #{defined_complexity.inspect} on #{field_defn.name}")
+          val = 0
+          @complexities_on_type_by_query.each do |query, complexities_on_type|
+            root_complexity = complexities_on_type.last
+            # Use this entry point to calculate the total complexity
+            total_complexity_for_query = ComplexityMergeFunctions.merged_max_complexity_for_scopes(query, [root_complexity.scoped_children])
+            val += total_complexity_for_query
           end
+          val
         end
 
-        # Selections on an object may apply differently depending on what is _actually_ returned by the resolve function.
-        # Find the maximum possible complexity among those combinations.
-        class AbstractTypeComplexity
-          def initialize
-            @types = Hash.new { |h, k| h[k] = {} }
-          end
-
-          # Return the max possible complexity for types in this selection
-          def max_possible_complexity
-            max = 0
-            @types.each_value do |fields|
-              complexity = fields.each_value.inject(:+)
-              max = complexity if complexity > max
+        # These functions use `ScopedTypeComplexity` objects,
+        # especially their `scoped_children`, to traverse down the tree
+        # and find the max complexity for any possible runtime type.
+        # Yowza.
+        module ComplexityMergeFunctions
+          module_function
+          # When looking at two selection scopes, figure out whether the selections on
+          # `right_scope` should be applied when analyzing `left_scope`.
+          # This is like the `isSubtypeOf` check from the GraphQL specification.
+          # TODO dedup with runtime checks? This is implemented somewhere.
+          def applies_to?(query, left_scope, right_scope)
+            if left_scope == right_scope
+              # This can happen when several branches are being analyzed together
+              true
+            else
+              # Check if these two scopes have _any_ types in common.
+              possible_right_types = query.possible_types(right_scope)
+              possible_left_types = query.possible_types(left_scope)
+              possible_right_types.each do |rt|
+                if possible_left_types.include?(rt)
```

Comment:
Could we use some built-in set operations here? `[...] & [...]`?

Reply:
Thanks for the suggestion! I _though_ my approach would be faster, since it could early-return and avoid allocating a result set, but I wrote a benchmark and found I was wrong: 

```ruby
require "benchmark/ips"

big_set_1 = (0..10).to_a
big_set_2 = (10..20).to_a

small_set_1 = (0..3).to_a
small_set_2 = (3..7).to_a

Benchmark.ips do |x|
  x.report("BIG include") {
    big_set_1.shuffle!
    big_set_2.shuffle!
    big_set_1.each do |i|
      next if big_set_2.include?(i)
    end
  }

  x.report("BIG [...] & [...] ") {
    big_set_1.shuffle!
    big_set_2.shuffle!
    !(big_set_1 & big_set_2).empty?
  }

  x.report("SMALL include") {
    small_set_1.shuffle!
    small_set_2.shuffle!
    small_set_1.each do |i|
      next if small_set_2.include?(i)
    end
  }

  x.report("SMALL [...] & [...] ") {
    small_set_1.shuffle!
    small_set_2.shuffle!
    !(small_set_1 & small_set_2).empty?
  }
end
```

```
~/code/graphql-ruby $ ruby inclusion_benchmark.rb
Warming up --------------------------------------
         BIG include    45.687k i/100ms
  BIG [...] & [...]     71.802k i/100ms
       SMALL include    98.258k i/100ms
SMALL [...] & [...]    156.726k i/100ms
Calculating -------------------------------------
         BIG include    466.952k (±18.5%) i/s -      2.239M in   5.037588s
  BIG [...] & [...]     845.262k (±12.1%) i/s -      4.236M in   5.116884s
       SMALL include      1.544M (± 6.2%) i/s -      7.762M in   5.054072s
SMALL [...] & [...]       2.316M (±18.5%) i/s -     11.128M in   5.037045s
```

I'll update it!